internal/marshaled/source.go

```diff
@@ -198,53 +211,92 @@ func (mds *DataSource) WatchItems(formatName string, iw source.ItemWatcher) erro
 
 func (mds *DataSource) startWatching() error {
 	// TODO: we could optimize the only-one-format-being-watched case
-	if !atomic.CompareAndSwapUint32(&mds.active, 0, 1) {
+	if mds.active {
 		return nil
 	}
+	mds.active = true
 	mds.itemChan = make(chan interface{}, mds.maxItems)
 	mds.itemsChan = make(chan []interface{}, mds.maxBatches)
-	go mds.processItemChan()
+	go mds.processItemChan(mds.itemChan, mds.itemsChan)
 	if mds.actiSource != nil {
 		mds.actiSource.Activate()
 	}
 	return nil
 }
 
 func (mds *DataSource) stopWatching() {
-	if !atomic.CompareAndSwapUint32(&mds.active, 1, 0) {
+	if !mds.active {
 		return
 	}
+	mds.active = false
 	for _, watcher := range mds.watchers {
 		watcher.Close()
 	}
 }
 
-func (mds *DataSource) processItemChan() {
-	for mds.Active() {
-		any := false
+func (mds *DataSource) processItemChan(itemChan chan interface{}, itemsChan chan []interface{}) {
+	stop := false
 
+loop:
+	for {
+		mds.watchLock.Lock()
+		active := mds.active
+		watchers := mds.watchers
+		mds.watchLock.Unlock()
+		if !active {
+			break loop
+		}
 		select {
-		case item := <-mds.itemChan:
-			for _, watcher := range mds.watchers {
+		case item, ok := <-itemChan:
+			if !ok {
+				itemChan = nil
+				continue loop
+			}
+			any := false
+			for _, watcher := range watchers {
```

Comment:
fair point, but I think I'm going to not start tugging at the "cleanup for readability" thread in this correctness PR ;-)